### DISCUSSION TITLE
Interaction between Supabase and Next.js' middleware + PPR

### DISCUSSION THREAD

**Discussion by chrisb2244** [supabase/supabase#21656] [Questions]
The supabase/ssr documentation instructs the developer to call `await supabase.auth.getUser()` in Next's `middleware`, in order to (as I understand it) ensure that for all pages the session/user is up-to-date (and that cookies are refreshed).

Middleware appears to run in all cases when PPR is enabled, so delays during middleware will contribute to a slower 'static' shell.

If I imagine a site with a user button ("Login/Register" if no user/session, "My Profile" if logged-in), and I want to render that on all pages (so I place it in the `layout.tsx`), then this part of the page will always require the Supabase calls to check state.
I can wrap this component in `Suspense` to allow the rest of the page to be loaded statically (assuming that for at least some pages, there exists a meaningful page for the anonymous user).

Does achieving a 'fast' response in these circumstances (mostly static page, dynamic element in Suspense, Supabase auth) require (and allow?) moving the `getUser()` call out of middleware and into some other element? (Ideally the header element, but apparently server components can't set cookies, so perhaps this requires an API route handler to update the user session?).

If this is the case (and I create some route `/api/supabase/updateUser/route.ts` which sets cookies via calling `getUser()`), when will the session/user be valid?
If I understand the rendering correctly, a layout like the following (typed, not copied, so forgive potential pseudo-code)
```
// layout.tsx (root)
...
return (
  <html>
  <head>...</head>
  <body>

  <Suspense fallback={<SomeStaticElement />}>
     <DynamicHeaderUsingSupabaseViaRouteHandler />
  </Suspense>

  {children}

  <StaticFooter />

  </body>
  </html>
);
```

will return the child page as if there was no user (i.e. the build-time page), and run the supabase elements then stream the icon/login button.

If the page should require login for any content, how should I then structure it?
```
// page.tsx requiring a user/session
const Page = async () => {
  // Probably this doesn't work because it can't set cookies?
  const user = (await supabase.auth.getUser()).data.user; // User | null, iirc.

  // Maybe this instead? (with disabling of cache either via POST method or force-dynamic or whatever is required by next.js)
  const user = await fetch('/api/my-supabase-route/update-user'); 

  if (user === null) {
    // What to do here? Redirect, return null, etc?
  }

  return (
    <div>My awesome logged in page</div>
  );
}

export default Page; 
// Maybe I can use a withAuth(Page) type handling here? 
// And define withAuth to manage the route handler call?
```

```
// Potential withAuth element?

// maybe the type here is a Promise, rather than a page? Not a supabase problem, anyway
// Callback may be invalid syntax, but imagining a default implementation 
// being return null, and allowing () => { redirect('/login') } or similar
export const withAuth = async (ChildPage: NextPage, nonAuthCallback: SomeCallbackType = () => null) => { 
  const user = ... // as above
  if (user === null) {
    return nonAuthCallback(); // or null?
  }
  return <ChildPage user={user} />;
}
```

Looking for any suggestions on the validity of these ideas, or general guidance on how Supabase will operate in a PPR-based system (where many pages are static and should ideally be returned without requiring a Supabase round-trip, but some require auth for specific pages or parts of pages).

**Comment by notjustinshaw** [comment]
Curious about this as well. Also, it seems like the latest canary version is now nextjs v15 which requires react v19. In the v15-rc blog post they mention that fetch calls are no longer cached by default.

I had a ppr test site that was using supabase before the v15 switch, and upgrading to v15 with the no-cache fetch calls started 429-ing the supabase server (I imagine because the server-side calls like getUser() from different server components used to be cached but now are not cached anymore).

I'd love to hear some thoughts from the team about the direction we're headed. Obviously this is bleeding-edge stuff here so I'm not expecting a timeline. Just curious how this will work.

**Reply by chrisb2244** [reply]
Regarding some of the caching, I think probably wrapping the getUser() or similar calls in React's `cache` function will be important (this doesn't cache between renders, but only within a single render pass, so handling the many components using the same function, deduplicating network calls).
You can see an example of this in Delba Oliveira's new Auth video here (with timestamp): [Next.js Authentication (YouTube)](https://youtu.be/N_sUsq_y10U?si=JwOBbswlGBt4NO0B&t=413)

For memoisation between renders, I think we'd need Next's `unstable_cache`, but I'm not sure if that's appropriate in the context of Supabase Auth (I'd guess probably not).
